File: projects/week_2/wizard_prompt_battle/wizard-battle-client/src/classes/gameState.js
import ActionType from "../enums/actionType";
import ActionTarget from "../enums/actionTarget";
import SpellType from "../enums/spellType";
import StatusEffectType from "../enums/statusEffectType";
import EffectGroup from "../enums/effectGroup";
import Player from "./player";
import PlayerState from "./playerState";
import StatusEffect from "./statusEffect";
import ActionRecord from "./actionRecord";

const shuffle = (array, randomFn = Math.random) => {
  const copy = [...array];
  for (let i = copy.length - 1; i > 0; i -= 1) {
    const j = Math.floor(randomFn() * (i + 1));
    [copy[i], copy[j]] = [copy[j], copy[i]];
  }
  return copy;
};

class GameState {
  constructor() {
    this.player_states = [];
    this.action_log = [];
  }

  initialize(wizard1, wizard2, randomFn = Math.random) {
    const assigned_order = shuffle([wizard1, wizard2], randomFn);
    this.player_states = assigned_order.map((wizard, index) => {
      const max_hp = wizard.max_hp(randomFn);
      return new PlayerState(
        new Player(index, wizard),
        max_hp,
        max_hp,
        wizard.starting_mana(randomFn),
        []
      );
    });

    this.action_log = [];
    return this.player_states.map((state) => state.player.wizard);
  }

  change_health(player, delta) {
    const state = this.player_states[player.id];
    state.current_health = Math.max(0, Math.min(state.max_health, state.current_health + delta));
    return state.current_health;
  }

  set_health(player, value) {
    const state = this.player_states[player.id];
    state.current_health = Math.max(0, Math.min(state.max_health, value));
    return state.current_health;
  }

  change_mana(player, delta) {
    const state = this.player_states[player.id];
    state.current_mana = Math.max(0, state.current_mana + delta);
    return state.current_mana;
  }

  set_mana(player, value) {
    const state = this.player_states[player.id];
    state.current_mana = Math.max(0, value);
    return state.current_mana;
  }

  add_status_effect(player, effect) {
    const state = this.player_states[player.id];
    const existing = state.active_effects.find((active) => active.name === effect.name);
    if (existing) {
      existing.remaining_turns = effect.remaining_turns;
      existing.value = effect.value;
      existing.effect_type = effect.effect_type;
      return;
    }
    state.active_effects.push(effect);
  }

  clear_expired_effects(player) {
    const state = this.player_states[player.id];
    state.active_effects = state.active_effects.filter((effect) => effect.remaining_turns > 0);
  }

  tick_effects() {
    this.player_states.forEach((state) => {
      state.active_effects.forEach((effect) => {
        effect.remaining_turns = Math.max(0, effect.remaining_turns - 1);
      });
    });
  }

  log_action(record) {
    this.action_log.push(record);
  }

  perform_action(actor_index, action, randomFn = Math.random) {
    if (actor_index !== 0 && actor_index !== 1) {
      throw new Error("actor_index must be 0 or 1");
    }

    const actor_state = this.player_states[actor_index];
    const defender_state = this.player_states[1 - actor_index];

    if (action.mana_cost() > actor_state.current_mana) {
      throw new Error("Not enough mana to perform action");
    }

    const result = action.perform_action(randomFn);

    if (!result?.succeeded) {
      this.log_action(new ActionRecord(actor_index, action.action_type, action.action_target(), "Failed :("));
      this._decay_effects(actor_index);
      return action.failure_announcement(actor_state.player.wizard);
    }

    actor_state.current_mana = Math.max(0, actor_state.current_mana - action.mana_cost());
    let final_action_value = result.value;

    switch (result.action_type) {
      case ActionType.HEAL: {
        const healed = this._apply_heal(actor_state, Number(result.value));
        final_action_value = healed;
        this.log_action(new ActionRecord(actor_index, ActionType.HEAL, ActionTarget.SELF, `Healed ${healed}`));
        break;
      }
      case ActionType.DEFEND: {
        this._apply_status(
          actor_state,
          new StatusEffect(
            action.element.name,
            StatusEffectType.DEFENSE,
            0.0,
            3
          )
        );
        this.log_action(new ActionRecord(actor_index, ActionType.DEFEND, ActionTarget.SELF, `Raised ${action.element.name} shield`));
        break;
      }
      case ActionType.CAST_SPELL: {
        switch (result.spell_type) {
          case SpellType.DAMAGE: {
            const damage = this._calculate_damage(actor_state, defender_state, action, Number(result.value));
            this._apply_damage(defender_state, damage);
            final_action_value = damage;
            this.log_action(new ActionRecord(actor_index, ActionType.CAST_SPELL, ActionTarget.ENEMY, `Dealt ${damage}`));
            break;
          }
          case SpellType.BUFF: {
            this._apply_status(
              actor_state,
              new StatusEffect(
                action.name,
                StatusEffectType.BUFF,
                Number(result.value),
                4
              )
            );
            this.log_action(new ActionRecord(actor_index, ActionType.CAST_SPELL, ActionTarget.SELF, `Buff ${action.name}`));
            break;
          }
          case SpellType.DEBUFF: {
            this._apply_status(
              defender_state,
              new StatusEffect(
                action.name,
                StatusEffectType.DEBUFF,
                Number(result.value),
                3
              )
            );
            this.log_action(new ActionRecord(actor_index, ActionType.CAST_SPELL, ActionTarget.ENEMY, `Debuff ${action.name}`));
            break;
          }
          default:
            throw new Error(`Unhandled spell type: ${result.spell_type}`);
        }
        break;
      }
      default:
        throw new Error(`Unhandled action type: ${result.action_type}`);
    }

    this._decay_effects(actor_index);

    return action.success_announcement(actor_state.player.wizard, final_action_value);
  }

  get_winner() {
    if (!this.player_states.length) {
      return null;
    }
    if (this.player_states[0].current_health <= 0) {
      return this.player_states[1].player.wizard;
    }
    if (this.player_states[1].current_health <= 0) {
      return this.player_states[0].player.wizard;
    }
    return null;
  }

  _decay_effects(actor_index) {
    const actor_state = this.player_states[actor_index];
    const defender_state = this.player_states[1 - actor_index];

    this._decrement_effects(actor_state, EffectGroup.BUFFS_AND_DEBUFFS);
    this._decrement_effects(defender_state, EffectGroup.DEFENSES);
  }

  _decrement_effects(state, group) {
    const updated = [];
    state.active_effects.forEach((effect) => {
      if (group.includes(effect)) {
        effect.remaining_turns = Math.max(0, effect.remaining_turns - 1);
      }
      if (effect.remaining_turns > 0) {
        updated.push(effect);
      }
    });
    state.active_effects = updated;
  }

  _apply_heal(state, amount) {
    const new_health = Math.min(state.max_health, state.current_health + amount);
    const healed = new_health - state.current_health;
    state.current_health = new_health;
    return healed;
  }

  _apply_status(state, effect) {
    const existing = state.active_effects.find((active) => active.name === effect.name);
    if (existing) {
      existing.remaining_turns = effect.remaining_turns;
      existing.value = effect.value;
      existing.effect_type = effect.effect_type;
      return;
    }
    state.active_effects.push(effect);
  }

  _calculate_damage(actor, defender, spell, base_damage) {
    let actor_multiplier = actor.player.wizard.damage_multiplier();
    actor.buffs().forEach((buff) => {
      actor_multiplier *= 1 + buff.value;
    });
    actor.debuffs().forEach((debuff) => {
      actor_multiplier *= Math.max(0, 1 - debuff.value);
    });

    let defender_multiplier = defender.player.wizard.damage_reduction();
    defender.buffs().forEach((buff) => {
      defender_multiplier *= 1 + buff.value;
    });
    defender.debuffs().forEach((debuff) => {
      defender_multiplier *= Math.max(0, 1 - debuff.value);
    });

    let damage = base_damage * actor_multiplier * defender_multiplier;

    defender.defenses().forEach((defense) => {
      const defense_element = defense.name;
      if (spell.element.strengths.includes(defense_element)) {
        damage *= 1.05;
      } else if (spell.element.weaknesses.includes(defense_element)) {
        damage *= 0.5;
      } else {
        damage *= 0.9;
      }
    });

    return Math.max(0, Math.round(damage));
  }

  _apply_damage(state, amount) {
    state.current_health = Math.max(0, state.current_health - amount);
  }

  increment_mana() {
    this.player_states.forEach((state) => {
      state.current_mana += state.player.wizard.mana_per_round();
    });
  }

  toString() {
    if (this.player_states.length < 2) {
      return "GameState: <uninitialized>";
    }

    const lines = this.player_states.map((state, idx) => `Player ${idx + 1}: ${state.toString()}`);
    lines.push("Actions:");
    if (!this.action_log.length) {
      lines.push("  (none)");
    } else {
      this.action_log.forEach((record, idx) => {
        const actor_label = record.actor_id === 0 ? "Player 1" : "Player 2";
        lines.push(`  ${idx + 1}. ${actor_label} -> ${record.type.name} (${record.target.name}) | ${record.result}`);
      });
    }

    return lines.join("\n");
  }
}

const game_state = new GameState();

export { GameState };
export default game_state;

File: projects/week_2/wizard_prompt_battle/game_state.py
"""
Central game state tracking for the wizard prompt battle.

This module exposes a single ``game_state`` instance that can be imported
anywhere in the project to inspect or mutate the current match. The data
structures capture both player snapshots and the action history so far.
"""

from __future__ import annotations

import random
from dataclasses import dataclass, field
from enum import Enum
from typing import Dict, List, Optional, Union

from classes import ActionTarget, ActionType, SpellType, Wizard, Action


class StatusEffectType(str, Enum):
    """Classifies an active effect."""

    DEFENSE = "defense"
    BUFF = "buff"
    DEBUFF = "debuff"


@dataclass
class StatusEffect:
    """Represents a single status effect on a player."""

    name: str
    effect_type: StatusEffectType
    value: float
    remaining_turns: int

    @property
    def is_buff(self) -> bool:
        return self.effect_type is StatusEffectType.BUFF

    @property
    def is_debuff(self) -> bool:
        return self.effect_type is StatusEffectType.DEBUFF

    @property
    def is_defense(self) -> bool:
        return self.effect_type is StatusEffectType.DEFENSE

    def __str__(self) -> str:
        return (
            f"{self.name}"
            f" [{self.effect_type}]"
            f" value={self.value}"
            f" turns={self.remaining_turns}"
        )


@dataclass
class Player:
    """Wraps a wizard with an assigned player identifier."""

    id: int
    wizard: Wizard


@dataclass
class PlayerState:
    """Snapshot of a player's current state."""

    player: Player
    max_health: int
    current_health: int
    current_mana: int
    active_effects: List[StatusEffect] = field(default_factory=list)

    def buffs(self) -> List[StatusEffect]:
        return [effect for effect in self.active_effects if effect.is_buff]

    def debuffs(self) -> List[StatusEffect]:
        return [effect for effect in self.active_effects if effect.is_debuff]

    def defenses(self) -> List[StatusEffect]:
        return [effect for effect in self.active_effects if effect.is_defense]

    def __str__(self) -> str:
        return (
            f"{self.player.wizard.name}: "
            f"HP {self.current_health}/{self.max_health}, "
            f"Mana {self.current_mana}, "
            f"Effects: {self._effects_summary()}"
        )

    def _effects_summary(self) -> str:
        if not self.active_effects:
            return "(none)"
        return ", ".join(str(effect) for effect in self.active_effects)


@dataclass
class ActionRecord:
    """Stores the outcome of a single action taken in the match."""

    actor_id: int  # 0 for player1, 1 for player2
    type: ActionType
    target: ActionTarget
    result: str


class EffectGroup(Enum):
    BUFFS_AND_DEBUFFS = 1
    DEFENSES = 2

    def includes(self, effect: StatusEffect) -> bool:
        if self is EffectGroup.BUFFS_AND_DEBUFFS:
            return effect.is_buff or effect.is_debuff
        if self is EffectGroup.DEFENSES:
            return effect.is_defense
        return False


class GameState:
    """Tracks the full state of an in-progress match."""

    def __init__(self) -> None:
        self.player_states: List[Player] = []
        self.action_log: List[ActionRecord] = []

    # ------------------------------------------------------------------
    # Initialization helpers
    # ------------------------------------------------------------------
    def initialize(
        self,
        wizard1: Wizard,
        wizard2: Wizard,
    ) -> List[Wizard]:
        """Reset the game state with fresh player snapshots."""

        assigned_order = [wizard1, wizard2]
        random.shuffle(assigned_order)

        self.player_states = []
        for index, wizard in enumerate(assigned_order):
            max_hp = wizard.max_hp()
            self.player_states.append(
                PlayerState(
                    player=Player(id=index, wizard=wizard),
                    max_health=max_hp,
                    current_health=max_hp,
                    current_mana=wizard.starting_mana(),
                )
            )

        self.action_log.clear()
        return [state.player.wizard for state in self.player_states]

    # ------------------------------------------------------------------
    # Mutators
    # ------------------------------------------------------------------
    def change_health(self, player: Player, delta: int) -> int:
        player_state = self.player_states[player.id]
        player_state.current_health = max(0, min(player_state.max_health, player_state.current_health + delta))
        return player_state.current_health

    def set_health(self, player: Player, value: int) -> int:
        player_state = self.player_states[player.id]
        player_state.current_health = max(0, min(player_state.max_health, value))
        return player_state.current_health

    def change_mana(self, player: Player, delta: int) -> int:
        player_state = self.player_states[player.id]
        player_state.current_mana = max(0, player_state.current_mana + delta)
        return player_state.current_mana

    def set_mana(self, player: Player, value: int) -> int:
        player_state = self.player_states[player.id]
        player_state.current_mana = max(0, value)
        return player_state.current_mana

    def add_status_effect(
        self,
        player: Player,
        effect: StatusEffect,
    ) -> None:
        player_state = self.player_states[player.id]
        for active_effect in player_state.active_effects:
            if active_effect.name == effect.name:
                active_effect.remaining_turns = effect.remaining_turns
                return

        player_state.active_effects.append(effect)

    def clear_expired_effects(self, player: Player) -> None:
        player_state = self.player_states[player.id]
        player_state.active_effects = [effect for effect in player_state.active_effects if effect.remaining_turns > 0]

    def tick_effects(self) -> None:
        """Reduce remaining turns for all active effects by one."""

        for player_state in self.player_states:
            for effect in player_state.active_effects:
                effect.remaining_turns = max(0, effect.remaining_turns - 1)

    def log_action(self, record: ActionRecord) -> None:
        self.action_log.append(record)

    def perform_action(self, actor_index: int, action: Action) -> Optional[dict]:
        if actor_index not in (0, 1):
            raise ValueError("actor_index must be 0 or 1")

        actor_state = self.player_states[actor_index]
        defender_state = self.player_states[1 - actor_index]

        if action.mana_cost() > actor_state.current_mana:
            raise ValueError("Not enough mana to perform action")

        result = action.perform_action()

        if not result.get("succeeded", False):
            self.log_action(ActionRecord(actor_index, action.action_type, action.action_target(), "Failed :("))
            self._decay_effects(actor_index)
            return action.failure_announcement(actor_state.player.wizard)

        actor_state.current_mana = max(0, actor_state.current_mana - action.mana_cost())

        final_action_value = result.get("value")

        match result.get("action_type"):
            case ActionType.HEAL:
                healed = self._apply_heal(actor_state, int(result["value"]))
                final_action_value = healed
                self.log_action(ActionRecord(actor_index, ActionType.HEAL, ActionTarget.SELF, f"Healed {healed}"))
            case ActionType.DEFEND:
                self._apply_status(
                    actor_state,
                    StatusEffect(
                        name=action.element.name,
                        effect_type=StatusEffectType.DEFENSE,
                        value=0.0,
                        remaining_turns=3,
                    ),
                )
                self.log_action(ActionRecord(actor_index, ActionType.DEFEND, ActionTarget.SELF, f"Raised {action.element.name} shield"))
            case ActionType.CAST_SPELL:
                spell_type = result.get("spell_type")
                match spell_type:
                    case SpellType.DAMAGE:
                        damage = self._calculate_damage(actor_state, defender_state, action, int(result["value"]))
                        self._apply_damage(defender_state, damage)
                        final_action_value = damage
                        self.log_action(ActionRecord(actor_index, ActionType.CAST_SPELL, ActionTarget.ENEMY, f"Dealt {damage}"))
                    case SpellType.BUFF:
                        self._apply_status(
                            actor_state,
                            StatusEffect(
                                name=action.name,
                                effect_type=StatusEffectType.BUFF,
                                value=float(result["value"]),
                                remaining_turns=4,
                            ),
                        )
                        self.log_action(ActionRecord(actor_index, ActionType.CAST_SPELL, ActionTarget.SELF, f"Buff {action.name}"))
                    case SpellType.DEBUFF:
                        self._apply_status(
                            defender_state,
                            StatusEffect(
                                name=action.name,
                                effect_type=StatusEffectType.DEBUFF,
                                value=float(result["value"]),
                                remaining_turns=3,
                            ),
                        )
                        self.log_action(ActionRecord(actor_index, ActionType.CAST_SPELL, ActionTarget.ENEMY, f"Debuff {action.name}"))

        self._decay_effects(actor_index)

        return action.success_announcement(actor_state.player.wizard, final_action_value)

    def get_winner(self) -> Optional[Wizard]:
        if self.player_states[0].current_health <= 0:
            return self.player_states[1].player.wizard
        if self.player_states[1].current_health <= 0:
            return self.player_states[0].player.wizard
        return None

    def _decay_effects(self, actor_index: int) -> None:
        actor_state = self.player_states[actor_index]
        defender_state = self.player_states[1 - actor_index]

        self._decrement_effects(actor_state, EffectGroup.BUFFS_AND_DEBUFFS)
        self._decrement_effects(defender_state, EffectGroup.DEFENSES)

    def _decrement_effects(self, state: PlayerState, group: EffectGroup) -> None:
        updated: List[StatusEffect] = []
        for effect in state.active_effects:
            if group.includes(effect):
                effect.remaining_turns = max(0, effect.remaining_turns - 1)
            if effect.remaining_turns > 0:
                updated.append(effect)
        state.active_effects = updated

    def _apply_heal(self, actor_state: PlayerState, amount: int) -> int:
        new_health = min(actor_state.max_health, actor_state.current_health + amount)
        healed = new_health - actor_state.current_health
        actor_state.current_health = new_health
        return healed

    def _apply_status(self, state: PlayerState, effect: StatusEffect) -> None:
        for existing in state.active_effects:
            if existing.name == effect.name:
                existing.remaining_turns = effect.remaining_turns
                existing.value = effect.value
                return
        state.active_effects.append(effect)

    def _calculate_damage(self, actor: PlayerState, defender: PlayerState, spell, base_damage: int) -> int:
        actor_multiplier = actor.player.wizard.damage_multiplier()
        for buff in actor.buffs():
            actor_multiplier *= (1 + buff.value)
        for debuff in actor.debuffs():
            actor_multiplier *= max(0.0, 1 - debuff.value)

        defender_multiplier = defender.player.wizard.damage_reduction()
        for buff in defender.buffs():
            defender_multiplier *= (1 + buff.value)
        for debuff in defender.debuffs():
            defender_multiplier *= max(0.0, 1 - debuff.value)

        damage = base_damage * actor_multiplier * defender_multiplier

        for defense in defender.defenses():
            defense_element = defense.name
            if defense_element in spell.element.strengths:
                damage *= 1.05
            elif defense_element in spell.element.weaknesses:
                damage *= 0.5
            else:
                damage *= 0.9

        return max(0, int(round(damage)))

    def _apply_damage(self, defender_state: PlayerState, damage: int) -> None:
        defender_state.current_health = max(0, defender_state.current_health - damage)

    def increment_mana(self) -> None:
        for state in self.player_states:
            state.current_mana += state.player.wizard.mana_per_round()

    # ------------------------------------------------------------------
    # Representation helpers
    # ------------------------------------------------------------------
    def __str__(self) -> str:
        if len(self.player_states) < 2:
            return "GameState: <uninitialized>"

        lines: List[str] = []
        for idx, state in enumerate(self.player_states, start=1):
            lines.append(f"Player {idx}: {state}")

        lines.append("Actions:")
        if not self.action_log:
            lines.append("  (none)")
        else:
            for idx, action_record in enumerate(self.action_log, start=1):
                actor_label = "Player 1" if action_record.actor_id == 0 else "Player 2"
                lines.append(
                    f"  {idx}. {actor_label} -> {action_record.type.name} ({action_record.target.name}) | {action_record.result}"
                )

        return "\n".join(lines)


# Expose a singleton-style instance that can be imported anywhere.
game_state = GameState()

__all__ = [
    "StatusEffectType",
    "StatusEffect",
    "PlayerState",
    "ActionRecord",
    "GameState",
    "game_state",
]

